src/main/java/org/opensearch/performanceanalyzer/commons/os/OSGlobals.java

```diff
@@ -67,20 +108,56 @@ private static void getScClkTckFromConfig() throws Exception {
         }
     }
 
+    private static void updateThreadStates() {
+        threadStates = ThreadList.getNativeTidMap(false);
+    }
+
+    private static boolean hasMatchedPattern(String threadName, List<Pattern> patterns) {
+        if (threadName == null) {
+            return false;
+        }
+        for (Pattern pattern : patterns) {
+            if (pattern.matcher(threadName).matches()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
     private static void enumTids() {
         tids.clear();
         tids.add(pid);
-
-        File self = new File("/proc/self/task");
-        File[] filesList = self.listFiles();
-        if (filesList != null) {
-            for (File f : filesList) {
-                if (f.isDirectory()) {
-                    String tid = f.getName();
-                    tids.add(tid);
+        // Maintaining a set to keep track of metrics not in either allowlist and denylist.
+        Set<String> otherThreadsMetrics = new HashSet<>();
+        updateThreadStates();
+        for (Map.Entry<Long, ThreadState> threadStateEntry : threadStates.entrySet()) {
+            Long tid = threadStateEntry.getKey();
+            ThreadState threadState = threadStateEntry.getValue();
+            // add this thread if the thread name is in allowlist
+            if (hasMatchedPattern(threadState.threadName, AllowedThreadNamePatterns)) {
+                tids.add(tid.toString());
+            }
+            // skip this thread if the thread name is in denylist
+            else if (hasMatchedPattern(threadState.threadName, DeniedThreadNamePatterns)) {
+                continue;
+            }
+            // if the thread name is neither in allowlist nor denylist, add
+            // this thread until certain threshold is reached.
+            else {
+                if (otherThreadsMetrics.size() < otherThreadsMetricsLimit) {
+                    otherThreadsMetrics.add(threadState.threadName);
+                    tids.add(tid.toString());
                 }
             }
         }
+        if (otherThreadsMetrics.size() >= otherThreadsMetricsLimit) {
+            StatsCollector.instance()
+                    .logException(StatExceptionCode.OTHER_THREAD_METRICS_NUMBER_EXCEEDS_THRESHOLD);
+        }
+        if (tids.size() >= totalThreadsMetricsLimit) {
+            StatsCollector.instance()
+                    .logException(StatExceptionCode.TOTAL_THREAD_METRICS_NUMBER_EXCEEDS_THRESHOLD);
+        }
```

Comment:
Let us add a metric to measure the time taken for filtering the threads and also a log message for audit here.